ipa-core/src/secret_sharing/mod.rs

```diff
@@ -57,13 +109,140 @@ pub trait Block: Sized + Copy + Debug {
+        Self: Vectorizable<1, Array = A>,
+        A: SharedValueArray<Self>,
+    {
+        once(self).collect::<A>()
+    }
+
+    fn from_array<A>(array: &A) -> Self
+    where
+        Self: Vectorizable<1, Array = A>,
+        A: SharedValueArray<Self>,
+    {
+        array.get(0)
+    }
+
+    fn from_array_mut<A>(array: &mut A) -> &mut Self
+    where
+        Self: Vectorizable<1, Array = A>,
+        A: SharedValueArray<Self>,
+    {
+        array.get_mut(0)
+    }
+}
+
+// Note that we can either make `trait Vectorizable<N>: SharedValue`, or we can make `trait
+// SharedValue: Vectorizable<1>`, but doing both creates a cycle. (Similarly for
+// `FieldVectorizable` / `Field`.)
+//
+// Although it is somewhat unnatural, we choose to do the latter, because it allows existing
+// high-level protocols unaware of vectorization to call vectorized versions of core protocols (with
+// width of 1) without updating all of the trait bounds. This does mean that the trait definitions
+// do not prevent implementing `Vectorizable` for something that is not a `SharedValue`, but please
+// don't do that.
+
+/// Trait for `SharedValue`s supporting operations on `N`-wide vectors.
+pub trait Vectorizable<const N: usize>: Sized {
+    type Array: SharedValueArray<Self>;
+}
+
+/// Trait for `Field`s supporting operations on `N`-wide vectors.
+///
+/// We would like `F` to be `FieldVectorizable` if it satisfies all of the following:
+///  1. `F: Field`.
+///  2. `<F as Vectorizable<N>>::Array: FieldArray<Self>`. Rust does not support expressing a
+///     constraint on a super-trait's associated type directly. Instead, this effect is achieved
+///     by constraining the `ArrayAlias` associated type and then constraining that
+///     `Vectorizable::Array == FieldVectorizable::ArrayAlias` where necessary (e.g. in the
+///     definition and blanket impl of the `FieldSimd` trait. We call it `ArrayAlias` instead of
+///     `Array` so that references to the `Array` associated type do not require qualification
+///     with a trait name.
+///  3. `F: Vectorizable<N>`. This is implied by the previous two, because `FieldArray`
+///     is a sub-trait of `SharedValueArray`.
+pub trait FieldVectorizable<const N: usize>: SharedValue + Sized {
+    type ArrayAlias: FieldArray<Self>;
+}
+
+// We could define a `SharedValueSimd` trait that is the analog of this for `SharedValue`s, but
+// there are not currently any protocols that need it.
+pub trait FieldSimd<const N: usize>:
+    Field + Vectorizable<N, Array = <Self as FieldVectorizable<N>>::ArrayAlias> + FieldVectorizable<N>
+{
+}
+
+// Portions of the implementation treat non-vectorized operations as a vector with `N = 1`. This
+// blanket impl (and the fact that `F: Field` is the only trait bound) is important in allowing code
+// that writes `F: Field` to continue working without modification.
+impl<F: Field> FieldSimd<1> for F {}
+
+// Supported vectorizations
+
+impl FieldSimd<32> for Fp32BitPrime {}
```

Comment:
clarifying question - is it for modulus conversion to be able to process all bits in parallel?

Reply:
We discussed today but to capture the answer, no, it's not to support any particular application, it's just the vectorization dimension I picked for the initial test cases in this PR. And, the vectorization dimension const parameter of `FieldSimd` being 32 is unrelated to the fact that it's `Fp32BitPrime` -- the dimension could be anything within reason.